### PR TITLE
Feature: FP16 stdlib support

### LINKED ISSUE #2293: Stdlib support for float16 type
Here's the list of missing function in stdlib for `float16` type:

- [ ] `uniform float16 extract(float16 x, uniform int i)`
- [ ] `float16 select(bool cond, float16 t, float16 f)`
- [ ] `float16 select(uniform bool cond, float16 t, float16 f)`
- [ ] `uniform float16 select(uniform bool cond, uniform float16 t, uniform float16 f)`
- [ ] `uniform float16 reduce_add(float16 x)`
- [ ] `uniform float16 reduce_min(float16 x)`
- [ ] `uniform float16 reduce_max(float16 x)`
- [ ] `uniform bool reduce_equal(float16 x)`
- [ ] `uniform bool reduce_equal(float16 x, uniform float16 *uniform value)`
- [ ] `float16 exclusive_scan_add(float16 v)`
- [ ] `void streaming_store(uniform float16 a[], float16 vals)`
- [ ] `void streaming_store(uniform float16 a[], uniform float16 vals)`
- [ ] `varying float16 streaming_load(uniform float16 a[])`
- [ ] `uniform float16 streaming_load_uniform(uniform float16 a[])`
- [ ] `int16 signbits(float16 x)`
- [ ] `unsigned int16 signbits(uniform float16 x)`
- [ ] `float16 round(float16 x)`
- [ ] `uniform float16 round(uniform float16 x)`
- [ ] `float16 floor(float16 x)`
- [ ] `uniform float16 floor(uniform float16 x)`
- [ ] `float16 ceil(float16 x)`
- [ ] `uniform float16 ceil(uniform float16 x)`
- [ ] `float16 trunc(float61 x)`
- [ ] `uniform float16 trunc(uniform float16 x)`
- [ ] `float16 sqrt(float16 v)`
- [ ] `uniform float16 sqrt(uniform float16 v)`
- [ ] `float16 rsqrt(float16 v)`
- [ ] `uniform float16 rsqrt(uniform float16 v)`
- [ ] `float16 ldexp(float16 x, int n)`
- [ ] `uniform float16 ldexp(uniform float16 x, uniform int n)`
- [ ] `float16 frexp(float16 x, varying int *uniform pw2)`
- [ ] `uniform float16 frexp(uniform float16 x, uniform int *uniform pw2)`

Trigonometric functions:

- [ ] `float16 sin(float16 x_full)`
- [ ] `float16 sin(uniform float16 x_full)`
- [ ] `float16 asin(float16 x_full)`
- [ ] `float16 asin(uniform float16 x_full)`
- [ ] `float16 cos(float16 x_full)`
- [ ] `float16 cos(uniform float16 x_full)`
- [ ] `float16 acos(float16 x_full)`
- [ ] `float16 acos(uniform float16 x_full)`
- [ ] `void sincos(float16 x_full, varying float16 *uniform sin_result, varying float16 *uniform cos_result)`
- [ ] `void sincos(uniform float16 x_full, uniform float16 *uniform sin_result, uniform float16 *uniform cos_result)`
- [ ] `float16 tan(float16 x_full)`
- [ ] `float16 tan(uniform float16 x_full)`
- [ ] `float16 atan(float16 x_full)`
- [ ] `float16 atan(uniform float16 x_full)`
- [ ] `float16 atan2(float16 y, float16 x)`
- [ ] `uniform float16 atan2(uniform float16 y, uniform float16 x)`

We don't have 16 bit atomics, so I put them optionally to decide on them later:
- [ ] Global atomics: `atomic_swap_global` / `atomic_compare_exchange_global`
- [ ] Local atomics: add, subtract, min, max, swap, compare_exchange

### PR DESCRIPTION
**Pull request by e3m3**
Closes #2293.

### CODE REVIEW

**Comment by e3m3** [comment]
Missing target implementations:
* [ ]   `__reduce_add_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                    * [ ]   `__reduce_min_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                    * [ ]   `__reduce_max_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                    * [ ]   `__reduce_equal_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                  * [ ]   `__exclusive_scan_add_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                            * [ ]   `__streaming_store_uniform_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                       * [ ]   `__streaming_store_varying_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                       * [ ]   `__streaming_load_uniform_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                        * [ ]   `__streaming_load_varying_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                        * [ ]   `__round_uniform_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                 * [ ]   `__round_varying_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                 * [ ]   `__floor_uniform_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                 * [ ]   `__floor_varying_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                 * [ ]   `__ceil_uniform_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                  * [ ]   `__ceil_varying_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                  * [ ]   `__trunc_uniform_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                 * [ ]   `__trunc_varying_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                 * [ ]   `__sqrt_uniform_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                  * [ ]   `__sqrt_varying_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                  * [ ]   `__rsqrt_uniform_half`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                 * [ ]   `__rsqrt_varying_half`.

File: stdlib.ispc [hunk 1]
```diff
@@ -2199,6 +2283,17 @@ static inline void *atomic_compare_exchange_local(void **ptr, void *oldval, void
 ///////////////////////////////////////////////////////////////////////////
 // Transcendentals (float precision)
 
+__declspec(safe) static inline float16 sqrt(float16 v) {
+    if (__math_lib == __math_lib_svml) {
+        const int16 iflt_neg_max = 0xff80; // -infinity
+        return halfbits(iflt_neg_max);     // FIXME: return __svml_sqrtf(v);
```
Comment (e3m3) [inline stdlib.ispc:2289]: How should we address `sqrt` for `__math_lib_svml`?

File: stdlib.ispc [hunk 2]
```diff
@@ -2209,6 +2304,17 @@ __declspec(safe) static inline float sqrt(float v) {
 
 __declspec(safe) static inline uniform float sqrt(uniform float v) { return __sqrt_uniform_float(v); }
 
+__declspec(safe) static inline float16 rsqrt(float16 v) {
+    if (__math_lib == __math_lib_svml) {
+        const int16 iflt_neg_max = 0xff80; // -infinity
+        return halfbits(iflt_neg_max);     // FIXME: return __svml_invsqrtf(v);
```
Comment (e3m3) [inline stdlib.ispc:2310]: How should we address `sqrt` for `__math_lib_svml`?

File: stdlib.ispc [hunk 3]
```diff
@@ -111,15 +111,9 @@ __declspec(safe, cost0) static inline uniform int __taskCount() { return __task_
 ///////////////////////////////////////////////////////////////////////////
 // Low level primitives
 
-__declspec(safe, cost0) static inline float16 float16bits(unsigned int16 a) { return __halfbits_varying_int16(a); }
+__declspec(safe, cost0) static inline float16 halfbits(unsigned int16 a) { return __halfbits_varying_int16(a); }
```
Comment (e3m3) [inline stdlib.ispc:114]: Renamed for consistency.

File: stdlib.ispc [hunk 4]
```diff
@@ -1083,6 +1122,7 @@ __declspec(safe) static inline uniform unsigned int64 reduce_max(unsigned int64
 REDUCE_EQUAL(int32, int32, IntMaskType)
 REDUCE_EQUAL(unsigned int32, int32, UIntMaskType)
 REDUCE_EQUAL(float, float, IntMaskType)
+REDUCE_EQUAL(float16, half, IntMaskType)
```
Comment (e3m3) [inline stdlib.ispc:1125]: Hope this is correct.

File: stdlib.ispc [hunk 5]
```diff
@@ -2243,6 +2369,26 @@ __declspec(safe) static inline uniform float ldexp(uniform float x, uniform int
...
+__declspec(safe) static inline uniform float16 frexp(uniform float16 x, uniform int *uniform pw2) {
+    uniform unsigned int16 ex = 0x7F80u;    // exponent mask
+    uniform unsigned int16 ix = intbits(x);
+    ex &= ix;
+    ix &= ~0x7F80u;                         // clear exponent
+    *pw2 = (uniform int)(ex >> 10) - 14;    // compute exponent
+    ix |= 0x3F00u;                          // insert exponent +1 in x
+    return halfbits(ix);
+}
+
```
Comment (e3m3) [inline stdlib.ispc:2391]: Check math.

File: stdlib.ispc [hunk 6]
```diff
@@ -2223,6 +2329,26 @@ __declspec(safe) static inline float rsqrt_fast(float v) { return __rsqrt_fast_v
...
+__declspec(safe) static inline uniform float16 ldexp(uniform float16 x, uniform int n) {
+    uniform unsigned int16 ex = 0x7F80u;
+    uniform unsigned int16 ix = intbits(x);
+    ex &= ix;               // extract old exponent;
+    ix = ix & ~0x7F800u;    // clear exponent
+    n = (n << 10) + ex;
+    ix |= n;                // insert new exponent
+    return halfbits(ix);
+}
```
Comment (e3m3) [inline stdlib.ispc:2350]: Check math.

**Comment by pema99** [comment]
Hey, was #2292 mentioned intentionally? It doesn't seem related

**Comment by e3m3** [comment]
> Hey, was #2292 mentioned intentionally? It doesn't seem related

Yes this is unrelated. This PR will also focus on stdlib support, with other related issues in separate PRs.

**Comment by pema99** [comment]
> > Hey, was #2292 mentioned intentionally? It doesn't seem related
> 
> Yes this is unrelated. This PR will also focus on stdlib support, with other related issues in separate PRs.

So just to be clear, #2292 is not fixed by this PR, but planned to be fixed in a future PR?

**Comment by e3m3** [comment]
> > > Hey, was #2292 mentioned intentionally? It doesn't seem related
> > 
> > 
> > Yes this is unrelated. This PR will also focus on stdlib support, with other related issues in separate PRs.
> 
> So just to be clear, #2292 is not fixed by this PR, but planned to be fixed in a future PR?

Correct.

File: stdlib.ispc (hunk 3 above)
Comment (dbabokin) [reply stdlib.ispc:114]: I don't think this is good idea. Current naming is after type names: `double`, `float`.

I also though that we should have a single `floatbits` function with overloaded parameters of type `int64`, `int32`, and `int16` - this way it's easier do meta-programming, as the name of function doesn't change. But I have doubts about "safety", as it's very easy to pass `int16` as `int32`.

File: stdlib.ispc [hunk 7]
```diff
@@ -941,15 +951,32 @@ __declspec(safe) static inline uniform unsigned int32 reduce_add(unsigned int16
...
+
 __declspec(safe) static inline uniform float reduce_add(float x) {
     // zero the lanes where the mask is off
     return __reduce_add_float(__mask ? x : 0.);
 }
 
+__declspec(safe) static inline uniform float16 reduce_min(float16 v) {
+    // For the lanes where the mask is off, replace the given value with
+    // infinity, so that it doesn't affect the result.
+    const int16 iflt_max = 0x7f80; // infinity
```
Comment (dbabokin) [inline stdlib.ispc:967]: Infinity for `float16` is `0x7c00`.

File: stdlib.ispc (hunk 4 above)
Comment (dbabokin) [reply stdlib.ispc:1125]: It's correct if and only if it's covered with tests that work ;-)

File: stdlib.ispc (hunk 1 above)
Comment (dbabokin) [reply stdlib.ispc:2289]: We need to add SVML function for half types and check what platforms are supported. I'll have a look at it later today.

File: stdlib.ispc [hunk 8]
```diff
@@ -2209,6 +2304,17 @@ __declspec(safe) static inline float sqrt(float v) {
 
 __declspec(safe) static inline uniform float sqrt(uniform float v) { return __sqrt_uniform_float(v); }
 
+__declspec(safe) static inline float16 rsqrt(float16 v) {
+    if (__math_lib == __math_lib_svml) {
+        const int16 iflt_neg_max = 0xff80; // -infinity
```
Comment (dbabokin) [inline stdlib.ispc:2309]: -Inf is `0xfc00` for float16.

File: stdlib.ispc [hunk 9]
```diff
@@ -2223,6 +2329,26 @@ __declspec(safe) static inline float rsqrt_fast(float v) { return __rsqrt_fast_v
 
 __declspec(safe) static inline uniform float rsqrt_fast(uniform float v) { return __rsqrt_fast_uniform_float(v); }
 
+__declspec(safe) static inline float16 ldexp(float16 x, int n) {
+    unsigned int16 ex = 0x7F80u;
```
Comment (dbabokin) [inline stdlib.ispc:2333]: these constants are incorrect for float16.

File: tests/abs-float16.ispc [hunk 10]
```diff
@@ -1,6 +1,5 @@
 #include "../test_static.isph"
-// rule: skip on arch=x86
-// rule: skip on arch=x86-64
+
```
Comment (dbabokin) [inline tests/abs-float16.ispc:2]: This PR doesn't have changes to make it work on x86, so we should probably not relax this restriction yet.

**Comment by dbabokin** [comment]
These changes were addressed by another PR #2307, closing this one

**Pull request closed without merging**